src/GraphQL/Extensions/GraphQLBuilderExtensions.cs

```diff
@@ -1168,5 +1171,52 @@ public static IGraphQLBuilder AddExecutionStrategy<TExecutionStrategy>(this IGra
         }
         #endregion
+
+        #region - UseTelemetry -
+#if NET5_0_OR_GREATER
+        /// <summary>
+        /// Configures the GraphQL server to collect traces via the <see cref="System.Diagnostics.Activity">System.Diagnostics.Activity API</see> and records events that match the
+        /// <see href="https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/graphql.md">OpenTelemetry recommendations</see>.
+        /// Trace data contains the GraphQL operation name, the operation type, and the optionally the document.
+        /// </summary>
+        /// <remarks>
+        /// When applicable, place after calls to UseAutomaticPersistedQueries to ensure that the query document is recorded properly.
+        /// </remarks>
+        public static IGraphQLBuilder UseTelemetry(this IGraphQLBuilder builder, ActivitySource activitySource, Action<TelemetryOptions>? configure = null)
+            => UseTelemetry(builder, _ => activitySource, configure != null ? (opts, _) => configure(opts) : null);
+
+        /// <inheritdoc cref="UseTelemetry(IGraphQLBuilder, ActivitySource, Action{TelemetryOptions}?)"/>
+        public static IGraphQLBuilder UseTelemetry(this IGraphQLBuilder builder, Func<IServiceProvider, ActivitySource> activitySourceFactory, Action<TelemetryOptions, IServiceProvider>? configure = null)
+        {
+            if (activitySourceFactory == null)
+                throw new ArgumentNullException(nameof(activitySourceFactory));
+
+            builder.Services.Configure(configure);
+            builder.ConfigureExecution(async (options, next) =>
+            {
+                var activitySource = activitySourceFactory(options.RequestServices ?? throw new MissingRequestServicesException());
+                using var activity = activitySource.StartActivity("graphql");
+                if (activity == null) // if no event listeners
+                    return await next(options).ConfigureAwait(false);
+                // record the requested operation name and optionally the GraphQL document
+                var telemetryOptions = options.RequestServices.GetRequiredService<TelemetryOptions>();
+                activity.SetTag("graphql.operation.name", options.OperationName);
+                if (telemetryOptions.RecordDocument)
+                    activity.SetTag("graphql.document", options.Query);
+                // record the operation type and the operation name (which may be specified within the
+                // document even if not specified in the request)
+                options.Listeners.Add(new TelemetryListener(activity));
+                // execute the request and set the status
+                var ret = await next(options).ConfigureAwait(false);
+#if NET6_0_OR_GREATER
+                var successful = !(ret.Errors?.Count > 0);
+                activity.SetStatus(successful ? ActivityStatusCode.Ok : ActivityStatusCode.Error);
```

Comment:
You should set the status on error only
https://opentelemetry.io/docs/concepts/signals/traces/#span-status

Reply:
Hmm... are you sure?  I read that document and also https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/README.md#setting-status and it doesn't seem to indicate to me that it should be set on error only.

Reply:
Note that the status might not be set if the request was cancelled prior to completion.  So it would seem that it would be valuable to have all three statuses (ok/error/unset) accurately represented in the telemetry.

Reply:
It says 

> When an exception is handled, a Span status can be set to Error. Otherwise, a Span status is in the Unset state.

Maybe @Kielek knows better.

Reply:
It seems `SetStatus` will work for [.NET6.0 but not 5.0](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/README.md#setting-status---diagnosticsource-version-older-than-60).
For .NET5.0 it should be
```csharp
activity.SetTag("otel.status_code", successful ? ActivityStatusCode.Ok : ActivityStatusCode.Error);
```

Reply:
That's an OpenTelemetry-specific tag and will not be included.  If you can find a reference to this on Microsoft's website, indicating that this is the recommended behavior, I'll be happy to include it.

Reply:
Makes sense. So maybe it's worth adding 
```csharp
public Action<Activity, ExecutionResult> EnrichWithExecutionResult { get; set; }
```
to the options so the users of the lower framework can configure consistent behavior.

Reply:
Similarly to `AspNetCoreInstrumentationOptions`
![image](https://user-images.githubusercontent.com/4518483/210113799-e2cb94a6-a310-4477-b174-43eed7bf6508.png)

Reply:
It's worth considering.  Keep in mind that the `IGraphQLBuilder` methods are not supposed to cover every single use case, but rather are supposed to cover the most common scenarios.

Some considerations:

- It may not be much code to for users to copy and paste the entire `UseTelemetry` implementation for their own needs
- .NET 5 is not a LTS release and is already unsupported by MS
- Anything prior to .NET 5 will require custom code to use with OpenTelemetry anyway
- We could make `TelemetryListener` public if a user requests it; but again, it's not much code and another goal is reducing the public API
- However, even users of .NET 6/7 may wish to tailor the `Activity` that is logged in some fashion; for example, by only adding documents that meet a filter criteria, or adding variables

Some methods of implementation:

- Add a single delegate to the options class to append more information than is done by default after the execution has completed (e.g. `EnrichWithExecutionResult`)
- Add three delegates to append info:
  - Initial
  - After document parsing
  - After completion
- Add three delegates which have default implementations, and so replacing any of them replaces default behavior
  - This is at odds with having options like `RecordDocument` set within options, since now that property may have no effect if the behavior is overridden
- Add some type of TelemetryWatcher class which has three overridable methods, and add `UseTelemetry<TWatcher>()`
- Extend the DocumentListener with hooks in the proper locations to implement entirely as a document listener, simplifying custom implementation -- people could derive from `TelemetryWatcher` and then call `AddDocumentListener<MyWatcher>()`
  - Note: now this is not possible because there is no way to hook into proper 'disposal' if an exception were to occur
- Leave as is presented here

Another consideration is that the API as presented in this PR should be fairly extendable to any of the above scenarios, should we choose to do so.

Besides leaving as-is, personally I favor extending the `DocumentListener` further.  However, this is a breaking change and would need to go into GraphQL.NET v8.

Second choice would be a delegate upon completion to append info.  Then we need to determine the proper signature.  What if execution was canceled (and so there is no `ExecutionResult`); does it run anyway?  Probably need `ExecutionOptions` as well as `ExecutionResult`, so services can be pulled from `ExecutionOptions` if need be, and for logging of variables, extensions, etc.

Reply:
Another idea: we could implement the `TelemetryWatcher` as a class that implements `IConfigureExecution`.  People can use `ConfigureExecution<MyTelemetryWatcher>()` if they want to use a derived class, at which point they can override any of the methods.  This makes sense and perhaps I will try this now, at least for the internal implementation.

Reply:
> Another idea: we could implement the `TelemetryWatcher` as a class that implements `IConfigureExecution`. People can use `ConfigureExecution<MyTelemetryWatcher>()` if they want to use a derived class, at which point they can override any of the methods. This makes sense and perhaps I will try this now, at least for the internal implementation.

Implemented this idea.

Reply:
> Similarly to `AspNetCoreInstrumentationOptions` ![image](https://user-images.githubusercontent.com/4518483/210113799-e2cb94a6-a310-4477-b174-43eed7bf6508.png)

Interesting.

Reply:
`otel.status_code`, plesae reference https://github.com/open-telemetry/opentelemetry-dotnet/issues/3045#issuecomment-1068314474

I would say that the best solution will be to reference disrectly System.Diganostics.DiagnosticsSource 7.0.0 and use always use SetStatus.

If it is not possible it will be great to have this attribute set as recommended (for otel purposes, it can be potentially ignored by other telemetry providers).

Reply:
Agree to add delegates like ones from `AspNetCoreInstrumentationOptions`.

> Add three delegates which have default implementations, and so replacing any of them replaces default behavior 

👍 

> This is at odds with having options like RecordDocument set within options, since now that property may have no effect if the behavior is overridden

Right. I suggested somewhere else in comments to switch from `RecordDocument` to `EnrichWithDocument`.

Reply:
Regarding setting `otel.status_code` explicitly from our code in case of pre-.NET6 platforms - I'm 50/50 here.

> for otel purposes, it can be potentially ignored by other telemetry providers

Maybe. Let's try?

Reply:
`otel.status_code` is explicitly suggested by Microsoft as an optional field for .NET 5, so if the consensus is to add the field for .NET 5, i'm okay with it.  .NET 5 is EOL already though, so I think it matters little.

https://learn.microsoft.com/en-us/dotnet/core/diagnostics/distributed-tracing-instrumentation-walkthroughs#optional-add-status

Reply:
> .NET 5 is EOL already though, so I think it matters little.

My use case - I have to use NET5 because of different security/certification reasons in constrained environments. EOL does not actually mean _end of use_. Many organizations will continue to use this TFM.

> if the consensus is to add the field for .NET 5, i'm okay with it.

Yes, let's add this tag with a little reminder comment about our discussion.